utilities/include/ray.hpp

```diff
@@ -0,0 +1,78 @@
+#ifndef RAY_HPP
+#define RAY_HPP
+
+#include <iostream>
+#include <cmath>
+#include "vec3.hpp"
+
+float epsilon=0.0001;
+class ray{
+    //origin
+    vec3 origin;
+
+    //unit vector denoting direction
+    vec3 direction;
+
+    public:
+        //Constructors
+        constexpr ray() : origin {vec3(0,0,0)}, direction {vec3(1,0,0)} {}
+        constexpr ray(vec3 Origin, vec3 Direction) : origin {Origin} , direction {Direction.unit()} {}
+
+        //fetch point at distance from point along ray, from origin by default
+        constexpr vec3 fetch(float distance, const vec3& point) const{
+            return point + distance * direction;
+        }
+
+        constexpr vec3 fetch(float distance) const{
+            return origin + distance * direction;
+        }
+
+        //origin setter
+        vec3& Origin(const vec3& Origin){
+            origin = Origin;
+            return origin;
+        }
+
+        //direction setter, maintains unit magnitude for direction
+        vec3& Direction(const vec3& Direction){
+            direction = Direction.unit();
+            return direction;
+        }
+
+        vec3 get_direction(){
+            return direction;
+        }
+
+        vec3 get_origin(){
+            return origin;
+        }
+        
+
+        //check for intersection between rays, dont use float equality
+        friend bool is_intersect(const ray& h, const ray& s);
```

Comment:
It might make more sense to name this function `is_intersecting` so that the code reads more like English, for example: `if (is_intersecting(ray1, ray2))` -> "if ray1 is intersecting with ray2".

Reply:
Also, is there any reason you named the variables `h` and `s` instead of, say, `ray1` and `ray2`?

Reply:
> Also, is there any reason you named the variables `h` and `s` instead of, say, `ray1` and `ray2`?

no, random characters, should I change them?

Reply:
Yes, you should always use reasonable variable names.